### PR TITLE
Execution service provider

### PR DESCRIPTION
**Pull request by hungneox**
+ Add ExecutionInterface and ExecutionProvider
+ Fix all of execution tests

### CODE REVIEW

File: src/Provider/ExecutionProvider.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    protected $provides = [
+        ExecutionInterface::class
+    ];
+
+    /**
+     * @inheritdoc
+     */
+    public function register()
+    {
+        $this->container->add(ExecutionInterface::class, function () {
```
Comment (crisu83) [inline src/Provider/ExecutionProvider.php]: Please change the second argument to `Execution::class`. Also, should this be a shared instance? If so, give `true` as the third argument.

File: src/facades.php [hunk 2]
```diff
@@ -101,7 +102,7 @@ function graphql(
     callable $fieldResolver = null
 ): ExecutionResult {
     /** @noinspection PhpParamsInspection */
-    return Execution::execute(
+    return executor()->execute(
```
Comment (crisu83) [inline src/facades.php]: I would prefer `GraphQL::get(ExecutionInterface::class)->execute(` here.

File: src/facades.php [hunk 3]
```diff
@@ -111,3 +112,11 @@ function graphql(
         $fieldResolver
     );
 }
+
+/**
+ * @return ExecutionInterface
+ */
+function executor(): ExecutionInterface
```
Comment (crisu83) [inline src/facades.php]: I think this method is unnecessary.

File: tests/Functional/Execution/ExecutionTest.php [hunk 4]
```diff
@@ -60,12 +62,12 @@ public function testNoSchemaIsProvided()
     {
         $this->expectException(\TypeError::class);
 
-        /** @var ExecutionResult $executionResult */
-        Execution::execute(null, parse('{field}'));
+        graphql(null, parse('{field}'));
```
Comment (crisu83) [inline tests/Functional/Execution/ExecutionTest.php]: Looks good! 👍

File: tests/Functional/Execution/ExecutionTest.php [hunk 5]
```diff
@@ -84,17 +86,20 @@ public function testAcceptAnObjectWithNamedPropertiesAsArguments()
...
         ]);
 
-        $rootValue    = 'rootValue';
-        $documentNode = parse('query Example { a }');
+        $rootValue = 'rootValue';
+        $document  = parse('query Example { a }');
 
         /** @var ExecutionResult $executionResult */
-        $result = Execution::execute($schema, $documentNode, $rootValue);
+        $result = executor()->execute($schema, $document, $rootValue);
```
Comment (crisu83) [inline tests/Functional/Execution/ExecutionTest.php]: I would prefer `$this->execution = GraphQL::get(ExecutionInterface::class);` inside the `setUp` method.

File: src/facades.php (hunk 3 above)
Comment (hungneox) [reply src/facades.php]: My test simply doesn't work with function `graphql(SchemaInterface $schema, string $source,...`. It looks unnecessary, but it's not. Let's I think a bit.

File: tests/Functional/Execution/ExecutionTest.php (hunk 4 above)
Comment (hungneox) [reply tests/Functional/Execution/ExecutionTest.php]: It's suppose not to work, the second parameter should be string . which is weird....

File: tests/Functional/Execution/ExecutionTest.php (hunk 5 above)
Comment (hungneox) [reply tests/Functional/Execution/ExecutionTest.php]: I agree, it's easier to change

File: src/facades.php (hunk 2 above)
Comment (hungneox) [reply src/facades.php]: Technically I can call only `graphql(...)` but for some reason all the tests failed.

File: src/helpers.php [hunk 6]
```diff
@@ -96,18 +97,56 @@ function graphql(
...
-    return Execution::execute(
-        $schema,
-        parse($source),
-        $rootValue,
-        $contextValue,
-        $variableValues,
-        $operationName,
-        $fieldResolver
-    );
+    return GraphQL::getInstance()
```
Comment (crisu83) [inline src/helpers.php]: This is supposed to be `GraphQL::get(...)`.

File: tests/Functional/Execution/ExecutionTest.php (hunk 4 above)
Comment (crisu83) [reply tests/Functional/Execution/ExecutionTest.php]: Shouldn't this be `graphql(null, '{field}')`?

File: tests/Functional/Execution/ExecutionTest.php [hunk 7]
```diff
@@ -84,17 +85,20 @@ public function testAcceptAnObjectWithNamedPropertiesAsArguments()
                 ])
         ]);
 
-        $rootValue    = 'rootValue';
-        $documentNode = parse('query Example { a }');
+        $rootValue = 'rootValue';
+        $document  = 'query Example { a }';
```
Comment (crisu83) [inline tests/Functional/Execution/ExecutionTest.php]: Ain't this `$source`?

File: src/helpers.php (hunk 6 above)
Comment (hungneox) [reply src/helpers.php]: Lol, I copied your code above...

File: tests/Functional/Execution/ExecutionTest.php (hunk 7 above)
Comment (hungneox) [reply tests/Functional/Execution/ExecutionTest.php]: It's DocumentNode, so that I call it document : D, but it's can be source.

**Pull request merged**